### PR TITLE
Support map as input/output type

### PR DESCRIPTION
**Pull request by honnix**
# TL;DR
Add native map support as input/output type.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Largely copied implementation of list, and fixed where `NotImplementedError` was raised.

## Tracking Issue
https://github.com/lyft/flyte/issues/637

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/common/types/containers.py [hunk 1]
```diff
@@ -149,3 +149,162 @@ def verbose_string(self):
...
+        # loading naturally interprets all layers, not just the outer layer.
+        return cls({k: cls.sub_type.from_python_std(v) for k, v in dictionary.items()})
+
+    @classmethod
+    def is_castable_from(cls, other):
+        """
+        :param flytekit.common.types.base_literal_types.FlyteSdkType other:
+        :rtype: bool
+        """
+        return type(cls) == type(other) and cls.sub_type.is_castable_from(other.sub_type)
```
Comment (honnix) [inline flytekit/common/types/containers.py:233]: I'm not entirely sure this is the right way to implement. I checked the code of `list` but it didn't pass any unit test I tried.

**Comment by honnix** [comment]
cc @kumare3

**Comment by kumare3** [comment]
@honnix do you mind waiting on the new flytekit api for adding map type support. It is already supported in the new api

**Comment by kumare3** [comment]
Here is the code for it
https://github.com/lyft/flytekit/blob/annotations/flytekit/annotated/type_engine.py#L303

And the usage
https://github.com/lyft/flytekit/blob/annotations/tests/flytekit/unit/annotated/test_type_hints.py#L948-L1024

**Comment by honnix** [comment]
@kumare3 Thanks for taking a look. Yes I took a look at the new type engine before and it looks a lot simpler and cleaner!

This is not that urgent because we have a workaround to skip returning a map as output. We would love to use the new API/SDK. What is the expected time to get out of alpha?

**Comment by kumare3** [comment]
@honnix alpha0 is out https://pypi.org/project/flytekit/0.16.0a0/. @wild-endeavor and I are working on releasing alpha1 which makes it almost feature complete (except for array/map tasks - different from dynamic tasks). We are hoping to have alpha1 early next week?

**Comment by honnix** [comment]
@kumare3 That sounds great. I guess it won't hurt that we start using (at least trying) alpha version.

BTW, is there any example showing how to use `FlyteContext` to set `auth_role`? Also how do we set `schedule` now when creating a launch plan (the factory method doesn't have such argument as I can see). I guess we can take this on Slack. Sorry for derailing. :)

**Comment by honnix** [comment]
There is a problem with the new API, as I see it requires Python 3.7+ due to `from __future__ import annotations`. Is that absolutely needed?

**Comment by wild-endeavor** [comment]
@honnix I thought I asked everyone that, sorry!  Can you not do 3.7?  We changed our minds a while back and decided to do 3.7 instead of just 3.6 and there's better typing support (even if hidden behind a future import).

**Comment by honnix** [comment]
@wild-endeavor Yeah I remember that conversion. It's not a big deal and we can prepare to get move to 3.7+. I will close this for now, unless we get an urgent need for it (things like we cannot wait for readiness of the new API, but unlikely to happen).

**Pull request closed without merging**